File: ApiAIDemoWatchOS/ApiAIDemoWatchOS WatchKit Extension/InterfaceController.m
/**
 * Copyright 2017 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
 
#import "InterfaceController.h"
#import <ApiAI/ApiAI.h>
#import <WatchConnectivity/WatchConnectivity.h>

@interface InterfaceController() <WCSessionDelegate>

@property(nonatomic, weak) IBOutlet WKInterfaceButton *button;
@property(nonatomic, weak) IBOutlet WKInterfaceImage *progressImage;

@end


@implementation InterfaceController

- (IBAction)doAction:(id)sender
{
    NSArray <WKAlertAction *> *actions =
  @[
    [WKAlertAction actionWithTitle:@"Text message"
                             style:WKAlertActionStyleDefault
                           handler:^{
                               [self sendTextRequest];
                           }],
    [WKAlertAction actionWithTitle:@"Cancel"
                             style:WKAlertActionStyleCancel
                           handler:^{
                               
                           }],
    ];
    
    [self presentAlertControllerWithTitle:@"Choose action"
                                  message:nil
                           preferredStyle:WKAlertControllerStyleActionSheet
                                  actions:actions];
}

- (void)sendTextRequest
{
    NSArray <NSString *> * suggestions = @[
                                           @"Hello",
                                           @"How are you?"
                                           ];
    [self presentTextInputControllerWithSuggestions:suggestions
                                   allowedInputMode:WKTextInputModePlain
                                         completion:^(NSArray * _Nullable results) {
                                             if (results.count) {
                                                 [self showProgress];
                                                 
                                                 ApiAI *apiai = [ApiAI sharedApiAI];
                                                 
                                                 AITextRequest *textRequest = [apiai textRequest];
                                                 
                                                 textRequest.query = @[results.firstObject];
                                                 
                                                 [textRequest setMappedCompletionBlockSuccess:^(AIRequest *request, AIResponse *response) {
                                                     NSString *text = response.result.fulfillment.speech;
                                                     
                                                     if (![text length]) {
                                                         text = @"<empty response>";
                                                     }
                                                     
                                                     [self.button setTitle:text];
                                                     
                                                     [self dismissProgress];
                                                 } failure:^(AIRequest *request, NSError *error) {
                                                     [self.button setTitle:[error localizedDescription]];
                                                     [self dismissProgress];
                                                 }];
                                                 
                                                 [apiai enqueue:textRequest];
                                             }
                                         }];

}

- (void)showProgress
{
    [_button setHidden:YES];
    [_progressImage setHidden:NO];
    [_progressImage startAnimating];
}

- (void)dismissProgress
{
    [_button setHidden:NO];
    [_progressImage setHidden:YES];
    [_progressImage stopAnimating];
}

- (void)willActivate {
    // This method is called when watch view controller is about to be visible to user
    [super willActivate];
}

- (void)didDeactivate {
    // This method is called when watch view controller is no longer visible
    [super didDeactivate];
}

@end

File: ApiAIDemoWatchOSSwift/ApiAIDemoWatchOSSwift WatchKit Extension/InterfaceController.swift
/**
 * Copyright 2017 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
 

import WatchKit
import Foundation
import ApiAI


class InterfaceController: WKInterfaceController {
    
    @IBOutlet weak var button: WKInterfaceButton!
    @IBOutlet weak var progressImage: WKInterfaceImage!
    
    @IBAction func doAction(_ sender: AnyObject) {
        let actions = [
            WKAlertAction(title: "Text Message", style: .default, handler: {[unowned self] () -> Void in
                self.textRequest()
            }),
            WKAlertAction(title: "Cancel", style: .cancel, handler: { () -> Void in
                // pass
            })
        ]
        
        self.presentAlert(
            withTitle: "Choose action", message:
            .none,
            preferredStyle: .actionSheet,
            actions: actions
        )
    }
    
    fileprivate func textRequest() {
        let sugessions = [
            "Hello",
            "How are your?"
        ]
        
        self.presentTextInputController(
            withSuggestions: sugessions,
            allowedInputMode: .plain) { (results) -> Void in
                guard let results = results as? [String] else {
                    return
                }
                
                guard let text = results.first else {
                    return
                }
                
                self.showProgress()
                
                let api = ApiAI.shared()
                
                let request = api?.textRequest()
                request?.query = text
                
                request?.setMappedCompletionBlockSuccess(
                    { (request, response) -> Void in
                        let response = response as! AIResponse
                        
                        var speech = response.result.fulfillment.speech
                        
                        if (speech?.characters.count == 0) {
                            speech = "<empty speech>"
                        }
                        
                        self.button.setTitle(speech)
                        
                        self.dismissProgress()
                    }, failure: { (request, error) -> Void in
                        self.dismissProgress()
                    }
                )
                
                api?.enqueue(request)
        }
    }
    
    fileprivate func showProgress() {
        button.setHidden(true)
        progressImage.setHidden(false)
        progressImage.startAnimating()
    }
    
    fileprivate func dismissProgress() {
        button.setHidden(false)
        progressImage.setHidden(true)
        progressImage.stopAnimating()
    }
}
